File: scripts/add_metadata_to_charts.py
#!/usr/bin/env python3
"""
Add PNG metadata to all chart files

This script updates all .qmd chart files to use add_png_metadata() for
proper attribution when images are shared on social media.
"""

import re
from pathlib import Path

CHART_FILES = [
    'dih-economic-models/figures/healthcare-spending-vs-life-expectancy.qmd',
    'dih-economic-models/figures/life-expectancy-pre-post-1962.qmd',
    'dih-economic-models/figures/fda-spending-life-expectancy-drug-costs-combined.qmd',
    'dih-economic-models/figures/life-expectancy-before-after-fda-line-chart.qmd',
    'dih-economic-models/figures/fda-spending-life-expectancy-trend-line-chart.qmd',
    'dih-economic-models/figures/fda-drug-development-cost-increase-line-chart.qmd',
    'dih-economic-models/figures/current-vs-dfda-clinical-trial-cost-column-chart.qmd',
    'dih-economic-models/figures/one-percent-treaty-top-country-contributions-column-chart.qmd',
    'dih-economic-models/figures/congressional-committee-seat-costs-column-chart.qmd',
    'dih-economic-models/figures/daily-deaths-breakdown-column-chart.qmd',
    'brain/book/appendix/peace-dividend-analysis.qmd',
    'brain/book/appendix/humanity-budget-overview.qmd',
    'dih-economic-models/figures/humanity-spending-priorities-bar-chart.qmd',
    'dih-economic-models/figures/war-vs-curing-diseases-column-chart.qmd',
    'dih-economic-models/figures/war-vs-disease-total-costs-vs-curing-spending-column-chart.qmd',
    'dih-economic-models/figures/war-hidden-direct-curing-diseases-column-chart.qmd',
    'dih-economic-models/figures/war-real-hidden-costs-breakdown.qmd',
    'dih-economic-models/figures/war-total-costs-breakdown-vs-curing-spending-column-chart.qmd',
    'dih-economic-models/figures/self-funding-roi-comparison-column-chart.qmd',
    'dih-economic-models/figures/military-vs-medical-research-spending-1-percent-treaty-column-chart.qmd',
    'dih-economic-models/figures/military-vs-medical-research-spending-column-chart.qmd',
    'dih-economic-models/figures/military-vs-medical-spending-ratio-bar-chart.qmd',
    'dih-economic-models/figures/money-flow-diagram.qmd',
    'dih-economic-models/figures/nih-budget-allocation-pie-chart.qmd',
    'dih-economic-models/figures/philanthropic-cost-effectiveness-comparison-bar-chart.qmd',
    'dih-economic-models/figures/public-health-interventions-economic-benefit-comparison-column-chart.qmd',
    'dih-economic-models/figures/military-vs-medical-research-direct-spending-column-chart.qmd',
    'dih-economic-models/figures/health-programs-vs-1-percent-treaty-societal-benefits-bar-chart.qmd',
    'dih-economic-models/figures/health-interventions-roi-comparison-column-chart.qmd',
    'dih-economic-models/figures/disease-war-curing-costs-column-chart.qmd',
    'dih-economic-models/figures/disease-vs-war-annual-deaths-pie-chart.qmd',
    'dih-economic-models/figures/disease-vs-curing-costs-column-chart.qmd',
    'dih-economic-models/figures/disease-burden-war-curing-comparison.qmd',
    'dih-economic-models/figures/dfda-investment-returns-bar-chart.qmd',
]


def generate_title_from_filename(filename: str) -> str:
    """Generate a title from a filename."""
    # Remove .png extension and convert hyphens to spaces
    name = filename.replace('.png', '').replace('-', ' ')
    # Capitalize each word
    return ' '.join(word.capitalize() for word in name.split())


def update_chart_file(filepath: Path) -> bool:
    """Update a single chart file with metadata. Returns True if modified."""
    content = filepath.read_text(encoding='utf-8')

    # Skip if already has add_png_metadata
    if 'add_png_metadata' in content:
        print(f'[OK] {filepath.name} - already updated')
        return False

    # Skip if no plt.savefig
    if 'plt.savefig' not in content:
        print(f'[SKIP] {filepath.name} - no plt.savefig found')
        return False

    original = content

    # 1. Add add_png_metadata to imports
    import_pattern = r'from brain\.figures\._chart_style import \(([\s\S]*?)\)'
    import_match = re.search(import_pattern, content)

    if import_match:
        imports = import_match.group(1)
        if 'add_png_metadata' not in imports:
            # Add to imports (on new line before closing paren)
            updated_imports = imports.rstrip() + ',\n    add_png_metadata'
            content = re.sub(
                import_pattern,
                f'from dih_economic_models.figures._chart_style import ({updated_imports})',
                content
            )

    # 2. Find and update plt.savefig calls
    # Pattern: plt.savefig(output_dir / 'filename.png', dpi=..., ...)
    savefig_pattern = r"plt\.savefig\(output_dir / '([^']+)'(.*?)\)"

    def replace_savefig(match):
        filename = match.group(1)
        args = match.group(2)
        title = generate_title_from_filename(filename)

        return f"""output_path = output_dir / '{filename}'

# Save the figure
plt.savefig(output_path{args})

# Add metadata for attribution
add_png_metadata(output_path, title="{title}")"""

    content = re.sub(savefig_pattern, replace_savefig, content)

    # Check if we made changes
    if content != original:
        filepath.write_text(content, encoding='utf-8')
        print(f'[UPDATED] {filepath.name}')
        return True
    else:
        print(f'[SKIP] {filepath.name} - no changes needed')
        return False


def main():
    """Process all chart files."""
    root = Path.cwd()
    print('Updating chart files with metadata...\n')

    updated_count = 0
    for file_path in CHART_FILES:
        full_path = root / file_path
        try:
            if full_path.exists():
                if update_chart_file(full_path):
                    updated_count += 1
            else:
                print(f'[ERROR] {file_path} - file not found')
        except Exception as e:
            print(f'[ERROR] {file_path} - error: {e}')

    print(f'\nDone! Updated {updated_count} files.')


if __name__ == '__main__':
    main()

File: scripts/add-metadata-to-charts.ts
#!/usr/bin/env tsx
/**
 * Add PNG metadata to all chart files
 *
 * This script updates all .qmd chart files to use add_png_metadata() for
 * proper attribution when images are shared on social media.
 */

import * as fs from 'fs';
import * as path from 'path';
import { saveFile } from './lib/file-utils';

const chartFiles = [
  'brain/figures/healthcare-spending-vs-life-expectancy.qmd',
  'brain/figures/life-expectancy-pre-post-1962.qmd',
  'brain/figures/fda-spending-life-expectancy-drug-costs-combined.qmd',
  'brain/figures/life-expectancy-before-after-fda-line-chart.qmd',
  'brain/figures/fda-spending-life-expectancy-trend-line-chart.qmd',
  'brain/figures/fda-drug-development-cost-increase-line-chart.qmd',
  'brain/figures/current-vs-dfda-clinical-trial-cost-column-chart.qmd',
  'brain/figures/one-percent-treaty-top-country-contributions-column-chart.qmd',
  'brain/figures/congressional-committee-seat-costs-column-chart.qmd',
  'brain/figures/daily-deaths-breakdown-column-chart.qmd',
  'brain/book/appendix/peace-dividend-analysis.qmd',
  'brain/book/appendix/humanity-budget-overview.qmd',
  'brain/figures/humanity-spending-priorities-bar-chart.qmd',
  'brain/figures/war-vs-curing-diseases-column-chart.qmd',
  'brain/figures/war-vs-disease-total-costs-vs-curing-spending-column-chart.qmd',
  'brain/figures/war-hidden-direct-curing-diseases-column-chart.qmd',
  'brain/figures/war-real-hidden-costs-breakdown.qmd',
  'brain/figures/war-total-costs-breakdown-vs-curing-spending-column-chart.qmd',
  'brain/figures/self-funding-roi-comparison-column-chart.qmd',
  'brain/figures/military-vs-medical-research-spending-1-percent-treaty-column-chart.qmd',
  'brain/figures/military-vs-medical-research-spending-column-chart.qmd',
  'brain/figures/military-vs-medical-spending-ratio-bar-chart.qmd',
  'brain/figures/money-flow-diagram.qmd',
  'brain/figures/nih-budget-allocation-pie-chart.qmd',
  'brain/figures/philanthropic-cost-effectiveness-comparison-bar-chart.qmd',
  'brain/figures/public-health-interventions-economic-benefit-comparison-column-chart.qmd',
  'brain/figures/military-vs-medical-research-direct-spending-column-chart.qmd',
  'brain/figures/health-programs-vs-1-percent-treaty-societal-benefits-bar-chart.qmd',
  'brain/figures/health-interventions-roi-comparison-column-chart.qmd',
  'brain/figures/disease-war-curing-costs-column-chart.qmd',
  'brain/figures/disease-vs-war-annual-deaths-pie-chart.qmd',
  'brain/figures/disease-vs-curing-costs-column-chart.qmd',
  'brain/figures/disease-burden-war-curing-comparison.qmd',
  'brain/figures/dfda-investment-returns-bar-chart.qmd',
];

async function updateChartFile(filePath: string): Promise<void> {
  const fullPath = path.join(process.cwd(), filePath);
  let content = fs.readFileSync(fullPath, 'utf-8');

  // Skip if already has add_png_metadata
  if (content.includes('add_png_metadata')) {
    console.log(`✓ ${filePath} - already updated`);
    return;
  }

  // Skip if no plt.savefig
  if (!content.includes('plt.savefig')) {
    console.log(`⊘ ${filePath} - no plt.savefig found`);
    return;
  }

  let modified = false;

  // 1. Add add_png_metadata to imports if not present
  const importRegex = /from brain\.figures\._chart_style import \(([\s\S]*?)\)/;
  const importMatch = content.match(importRegex);

  if (importMatch) {
    const imports = importMatch[1];
    if (!imports.includes('add_png_metadata')) {
      // Add to imports
      const updatedImports = imports.trim() + ',\n    add_png_metadata';
      content = content.replace(importRegex, `from brain.figures._chart_style import (${updatedImports})`);
      modified = true;
    }
  }

  // 2. Update plt.savefig calls
  // Match patterns like:
  // plt.savefig(output_dir / 'filename.png', ...)
  // plt.savefig(some_path, ...)
  const savefigRegex = /plt\.savefig\(([\s\S]*?)\)/g;
  let match;
  const updates: Array<{ original: string; replacement: string }> = [];

  while ((match = savefigRegex.exec(content)) !== null) {
    const fullCall = match[0];
    const args = match[1];

    // Extract the file path (first argument)
    const pathMatch = args.match(/^([^,]+)/);
    if (!pathMatch) continue;

    const pathExpr = pathMatch[1].trim();

    // Generate title from filename
    const filenameMatch = pathExpr.match(/['"]([^'"]+\.png)['"]/);
    let title = '';
    if (filenameMatch) {
      const filename = filenameMatch[1].replace('.png', '').replace(/-/g, ' ');
      title = filename
        .split(' ')
        .map(word => word.charAt(0).toUpperCase() + word.slice(1))
        .join(' ');
    }

    // Check if we need to store path in a variable
    const needsVariable = !pathExpr.includes('output_path') && !pathExpr.match(/^[a-z_]+$/);

    let replacement = '';
    if (needsVariable) {
      // Need to create output_path variable
      replacement = `output_path = ${pathExpr}\nplt.savefig(output_path`;
      // Keep the rest of the arguments
      const restOfArgs = args.substring(pathMatch[0].length);
      replacement += restOfArgs + ')';

      // Add metadata call
      replacement += `\nadd_png_metadata(output_path, title="${title}")`;
    } else {
      // Path is already a variable
      const varName = pathExpr;
      replacement = fullCall + `\nadd_png_metadata(${varName}, title="${title}")`;
    }

    updates.push({ original: fullCall, replacement });
  }

  // Apply updates in reverse order to preserve positions
  for (const update of updates.reverse()) {
    content = content.replace(update.original, update.replacement);
    modified = true;
  }

  if (modified) {
    await saveFile(fullPath, content);
    console.log(`✓ ${filePath} - updated`);
  } else {
    console.log(`⊘ ${filePath} - no changes needed`);
  }
}

// Process all files
async function main() {
  console.log('Updating chart files with metadata...\n');
  for (const file of chartFiles) {
    try {
      await updateChartFile(file);
    } catch (error) {
      console.error(`✗ ${file} - error: ${error}`);
    }
  }
  console.log('\nDone!');
}

main().catch(console.error);
